src/pycea/tl/permutation.py

```diff
@@ -0,0 +1,383 @@
+        Callable function that can reduce the data from all the leaves of a given split to a vector or scalar. Defaults
+        to np.mean(,axis = 0). Only used for test="permutation".
+    metric
+        A metric to compare the children from both sides of the tree. Can be a known metric or a callable. Only used
+        for test="permutation".
+    metric_kwds
+        Options for the metric.
+    test
+        Type of test to perform to compare the two groups. "t-test" can only be used for scalar quantities.
+    equal_var
+        Boolean indicating if the variance in the two groups should be assumed to be equal. Only used for
+        test="t-test".
+    n_permutations
+        Upper bound on the number of permutations to run. The actually executed
+        number is ``min(n_permutations, comb(n_left + n_right, n_left))`` per split.
+    min_group_leaves
+        Minimum number of leaves required in each group to perform a statistical test. The t-test may be particularly
+        unreliable with small sample sizes.
+    keys_added
+        Attribute keys of `tdata.obst[tree].nodes` where split statistics will be stored. If `None`, `keys` are used.
+    tree
+        The `obst` key or keys of the trees to use. If `None`, all trees are used.
+    copy
+        If True, returns a :class:`DataFrame <pandas.DataFrame>` with split statistics.
+
+    Returns
+    -------
+    pd.DataFrame or None
+        If ``copy=True``, a DataFrame indexed by node identifiers with columns
+        ``f"{key_added}_split"`` and ``f"{key_added}_pval"`` for each requested key.
+        Nodes where the permutation test was skipped will have missing columns/values
+        for that key. If ``copy=False``, returns ``None`` after writing attributes
+        onto the graphs.
+    """
+
+    if isinstance(keys, str):
+        keys = [keys]
+    if keys_added is None:
+        keys_added = keys
+    if isinstance(keys_added, str):
+        keys_added = [keys_added]
+    if len(keys) != len(keys_added):
+        raise ValueError("Length of keys must match length of keys_added.")
+    tree_keys = tree
+    _check_tree_overlap(tdata, tree_keys)
+    trees = get_trees(tdata, tree_keys)
+
+    if metric == "mean_difference":
+        metric_fn = MeanDiffMetric()
+    else:
+        metric_fn = DistanceMetric.get_metric(metric, **(metric_kwds or {}))
+
+    aggregate_fn = Aggregator.get_aggregator(aggregate)
+    df_dict = {}
+
+    first_key = True
+
+    for key, key_added in zip(keys, keys_added):
+
+        # lists for dataframe if copy = True
+        parent_list = []
+        group1_list = []
+        group2_list = []
+        group1_value_list = []
+        group2_value_list = []
+        pvalue_list = []
+        tree_list = []
+
+        data, is_array, is_square = get_keyed_obs_data(tdata, key)
+        if is_array or is_square and test == "t-test":
+            raise ValueError("t-test cannot be performed for vector valued keys.")
+        data = data.dropna()
```

Comment:
Need to do this after ```data = data[key]```. Otherwise it will drop any row that contains an NA value.